[Conrad Lew] IP

ConradLew opened the pull request:
## Duke
> “Your mind is for having ideas, not holding them.” – David Allen ([source](https://dansilvestre.com/productivity-quotes))

Duke frees your mind of having to remember things you need to do. It's,
- text-based
- easy to learn
- ~~FAST~~ SUPER FAST to use

All you need to do is,
1. download it from [here](https://nus-cs2103-ay2223s1.github.io/website/schedule/week4/project.html).
2. double-click it.
3. add your tasks.
4. let it manage your tasks for you :wink:

And it is **FREE!**

Features:
- [ ] Managing tasks
- [ ] Managing deadlines
- [ ] Reminders
---
If you are a Java programmer, you can use it to practice Java too. Here's the `main` method:

```Java
public class Main {
    public static void main(String[] args) {
        Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

ConradLew committed:
Added features for level 1

ConradLew committed:
Added features for Level 2

ConradLew committed:
Added features for Level 3

ConradLew committed:
Added features for Level 4

ConradLew committed:
Added automated test testing for A-TextUiTesting

ConradLew committed:
Added functionality for level-5

ConradLew committed:
Added functionality for Level-6, added more error handling for Level-5

ConradLew committed:
Added enum structure for A-Enums, however unapplicable to program

ConradLew committed:
Added basic javadocs to all existing files

ConradLew committed:
Add level-7 functionalities

ConradLew committed:
Add level-8 functionalities

ConradLew committed:
Fix merge conflicts in Duke.java

ConradLew committed:
Add more OOP principles to Duke

ConradLew committed:
Add A-MoreOOP functionality

ConradLew committed:
Organise classes into a single package duke

ConradLew committed:
Merge branch 'add-gradle-support'

ConradLew committed:
Add JUnit Tests for Deadline and Parser classes

ConradLew committed:
Package app as JAR file

ConradLew committed:
Add Javadocs to the code

ConradLew committed:
Revert "Package app as JAR file"

This reverts commit 387d393912e5c131664907b71adb684c4ea6059f.

ConradLew committed:
Revert "Revert "Package app as JAR file""

This reverts commit 1bb5003c902cfff2070693d571f13a18f25d54ba.

ConradLew committed:
Refactor code to be in line with module coding standard

ConradLew committed:
Add Level-9 functionalities

ConradLew committed:
Fix merge conflicts between Level-9 and A-CodingStandard

ConradLew committed:
Change Storage.java to account for missing local file

ConradLew committed:
Add checkstyle and updated code to be in line with coding standards

ConradLew committed:
Add UI for chat bot

ConradLew committed:
Add varargs for adding of multiple tasks

ConradLew committed:
Change javadocs to fit coding standard

zsiggg commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,87 @@
+    private Ui ui;
+
+    /**
+     * Constructor for the <code>Duke</code> chatbot.
```
Not sure if there is an error here, for \<code>Duke\</code>?

zsiggg commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,119 @@
+    public String showLoadingError(Exception e) {
+        return e.getMessage();
+    }
+
```
`showLoadingError` has the same implementation as `showError`. Perhaps could consider using just one method out of the two?

zsiggg commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,87 @@
+            tasks = new TaskList();
+        } catch (IOException e) {
+            ui.showLoadingError(e);
+        }
```
I'm assuming that `IOException` would be thrown from `storage.loadTasks()`, due to an error when reading the output file. Will there be a need to reinitialise `tasks`, since this error would mean that there are no initial tasks to be shown? 

To illustrate,
```java
} catch (IOException e) {
    ui.showLoadingError(e);
    tasks = new TaskList();
}
```

zsiggg commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,87 @@
+            case FIND:
+                return this.ui.showMatchingTasks(Parser.parseSearchInput(userResponse), this.tasks);
+            default:
+                break;
```
Would there be an error message shown if the user gives an invalid command?

zsiggg commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,172 @@
+ * @author Conrad
+ */
+public class Parser {
+
```
Perhaps we should make `Parser` abstract, since it only contains static methods?

zsiggg commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,172 @@
+        String[] parsedUserResponse = userInput.split(" ");
+        String newTaskDescription = String.join(" ",
+                Arrays.copyOfRange(parsedUserResponse, 1, parsedUserResponse.length));
+        return new Todo(newTaskDescription);
```
Maybe we could do something like this, assuming we are trying to obtain the todo description from the command given?
```java
String[] newTaskDescription = userInput.replace("/todo", "").strip();
```

zsiggg commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,172 @@
+                        parsedUserResponse.length));
+        return new Event(newTaskDescription, newEventTime);
+    }
+
```
Using the same idea as the comment on `parseTodoTask`, and assuming we are trying to get the description (which is before /at) and the time (which is after /at) maybe we could do this? If you agree it is easier to understand.
```java
String[] parsedUserResponse = userInput.replace("/event").split("/at");
String newTaskDescription = parsedUserResponse[0];
String newEventTime = parsedUserResponse[1];
```

zsiggg reviewed (commented):
Really amazing that you used JavaFX to create a GUI 😃. Just got a few things that I feel could make the code cleaner :)

Kennethsim2000 commented on src/test/java/duke/DeadlineTest.java:
```diff
@@ -0,0 +1,22 @@
+
+    @Test
+    public void testGetTextRepresentation() {
+        Deadline testDeadline = new Deadline("test", LocalDate.parse("1999-02-02"));
```
Indentation for wrapped line should be 8 spaces

Kennethsim2000 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,119 @@
+     */
+    public String showMatchingTasks(String searchInput, TaskList tasks) {
+        String matchingStart = "Here are the matching tasks in your list:\n";
+        return matchingStart + tasks.getMatchingTasksRepresentation(searchInput);
```
indentation for wrapped lines should be 8 spaces. E.g.
setText("Long line split"
        + "into two parts.");

Kennethsim2000 reviewed (commented):
Overally, good job with the code. LGTM:)

ConradLew committed:
Add assertions to project

ConradLew committed:
Add the ability to detect duplicate tasks.

Duke chatbot is unable to detect duplicate tasks.

An automatic checker for duplicate tasks is more user-friendly
because users cannot be expected to remember what tasks they have
already added to the list.

Let's,
* Update the task adding algorithm in the TaskList class to check for
duplicate tasks
* Add an error message if the user attempts to add duplicate tasks

Modifying the TaskList class is preferable as it follows the single
responsibility principle since this class is responsible for the
handling of tasks in the list.

ConradLew committed:
Replace Command enum for abstract class

Enum is used for encapsulating the different commands.

The common behaviours across different commands cause code duplication.

Extracting the common behaviour into an abstract class allows for the
use of polymorphism when running a user command in the bot. For example,
when the bot is executing a command, it does not have to check what
command it is before executing. This reduces the lengthiness of the
getResponse method in the Duke class.

Let's pull up behaviours common to all command classes into a new
abstract parent class called Command.

Using inheritance is preferable over composition in this situation
because the common behaviors are not composable.

ConradLew committed:
Remove repeated method in the Ui class

ConradLew committed:
Merge pull request #1 from ConradLew/branch-A-CodeQuality

Branch a code quality

ConradLew committed:
Merge branch 'master' into branch-A-Assertions

ConradLew committed:
Merge pull request #2 from ConradLew/branch-A-Assertions

Branch a assertions

ConradLew committed:
Merge branch 'master' into branch-C-DetectDuplicates

ConradLew committed:
Merge pull request #3 from ConradLew/branch-C-DetectDuplicates

Branch c detect duplicates

ConradLew committed:
Add try-catch block to execute statements

ConradLew committed:
Add streams to process duplicate tasks

ConradLew committed:
Set up CI for project

ConradLew committed:
Remove text IO redirection from CI

ConradLew committed:
Improve GUI for bot

ConradLew committed:
Fix issue #4 by using a more boolean sounding variable

ConradLew committed:
Improve error handling and messages to be more accurate

ConradLew committed:
Update user guide

ConradLew committed:
Update user guide

ConradLew committed:
Update user guide

ConradLew committed:
Fix relative paths in storage

ConradLew committed:
Fix formatting bug in JUnit test

ConradLew committed:
Update javadocs to contain latest features

ConradLew committed:
Replace buggy image in UI and fix resizing
